### DISCUSSION TITLE
growatt noah 2000 battery with other brand inverter

### DISCUSSION THREAD

**Discussion by cgatnet** [johanmeijer/grott#560] [Q&A]
Hi, I have a hoymiles inverter which is working fine and added a growatt Noah 2000.
Now I would like to read "solar input power" and "battery level" from it.
Is it possible to use grott for this environment?
I tried, but was not successful so far. Grott does start, but there is no visible com.
I reconfigured the "mqtt.growatt.com" (default in noah 2000 wifi setup) server to grott and also the port "7006". 
Maybe some hint or ideas? :)
Best regards, Christian

**Comment by cgatnet** [comment]
interesting: the noah 2000 is still online for the server, so... this is strange.

**Reply by cgatnet** [reply]
I guess, it has a fallback to default server, if com is failing with proxy / grott

**Comment by johanmeijer** [comment]
This a new device for me. We have to found out if the data protocol complies with other Growatt devices and Grott can read it. 

Did you install Grott and have be able to route the data via Grott? Then maybe we can found out from the log message if it is working or not.

**Reply by cgatnet** [reply]
Yes, installed it and have difficulties with rerouting. Only one time I got a response in the grott log.

**Reply by cgatnet** [reply]
Grott proxy mode started
	 **********************************************************************************
	 - Grott - libscrc not installed, no CRC checking only record validation on length!
	 **********************************************************************************
Hostname : HomeServer2
IP :  127.0.1.1 , port :  7006 

	 - ('192.168.2.177', 56662) has connected

	 - Growatt packet received:
		  <socket.socket fd=4, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('192.168.2.123', 57370), raddr=('47.254.130.145', 7006)>
	 - Grott - grottproxy - Invalid data record received, processing stopped for this record
	 - ('47.254.130.145', 7006) has disconnected
	 - ('192.168.2.177', 56663) has connected

	 - Growatt packet received:
		  <socket.socket fd=4, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('192.168.2.123', 45120), raddr=('8.209.71.240', 7006)>
	 - Grott - grottproxy - Invalid data record received, processing stopped for this record

**Comment by johanmeijer** [comment]
Ok. Grott does not recognize this as a valid Growatt data record. So it is different then the standard inverter (and battery connected via the inverter) records. 

This will be a challenge to decode I think (we need a protocol description). 

I also see that I do not display the content of the record in the debug messages. I have to change grott for this. 

Which inverter do you use (Growatt NEO?) and how communicate this with growatt (is there wifi integrated in the inverter?). 

By the way interesting device. Does it works ? Can you configure this via the growatt site (e.g when to charge/discharge?).

**Reply by cgatnet** [reply]
Yes, I can schedule times for different output rates and have the option to only charge the battery or do output first, so only solar "overpower" is charging the battery. If I would be able to read the battery level, it would be top rated! ;o)

**Reply by cgatnet** [reply]
actually I change the output of the connected "Hoymiles 800W" inverter, so I do not loose power to the energy provider, who does not pay me for it. :)

**Reply by cgatnet** [reply]
Its a Hoymiles "HM-800" inverter with a special protocol handled by this project here: [Ahoy](https://github.com/lumapu/ahoy)
So, no growatt inverter on my side.

**Comment by cgatnet** [comment]
@johanmeijer I quickly added this to your code:
`print(f"\t - Grott - grottproxy - Invalid data record received, processing stopped for this record")
print("data:"+vdata)`
and got this:
`data:16030100ef010000eb0303669f9cb7c07b935e9f98864dbca8605b60174134c655c0200a02f9f09678922e000062c02cc030009fc0adc09fc024c028006bc00ac0140039c0afc0a3c02bc02f009ec0acc09ec023c0270067c009c0130033c0aec0a2009dc09d003d0035c032c02ac00fc02ec026c005c0a1009cc09c003c002fc031c029c00ec02dc025c004c0a000ff0100006000000012001000000d3139322e3136382e322e313233000d001600140603060105030501040304010303030102030201000a001a00180019001c0018001b00170016001a0015001400130012001d000b00020100001600000017000000230000`
maybe you can read in the "Matrix"? ;o)

**Reply by cgatnet** [reply]
additionally I got 2 more, then the replies stopped again, unsure why.
`data:16030100ef010000eb0303669f9d11e50e5c503e5b249ac40a4c15eab8d7e588420609a4d36478f25c35b7000062c02cc030009fc0adc09fc024c028006bc00ac0140039c0afc0a3c02bc02f009ec0acc09ec023c0270067c009c0130033c0aec0a2009dc09d003d0035c032c02ac00fc02ec026c005c0a1009cc09c003c002fc031c029c00ec02dc025c004c0a000ff0100006000000012001000000d3139322e3136382e322e313233000d001600140603060105030501040304010303030102030201000a001a00180019001c0018001b00170016001a0015001400130012001d000b00020100001600000017000000230000`
`data:16030100ef010000eb0303669f9d6c678f2130e1de3b25ab3b41d8b40ff3813500fbdbb453f39e6519b886000062c02cc030009fc0adc09fc024c028006bc00ac0140039c0afc0a3c02bc02f009ec0acc09ec023c0270067c009c0130033c0aec0a2009dc09d003d0035c032c02ac00fc02ec026c005c0a1009cc09c003c002fc031c029c00ec02dc025c004c0a000ff0100006000000012001000000d3139322e3136382e322e313233000d001600140603060105030501040304010303030102030201000a001a00180019001c0018001b00170016001a0015001400130012001d000b00020100001600000017000000230000`

**Comment by johanmeijer** [comment]
Honestly. I can mot make any chocolate of this at this moment. It looks like this record does not has the same layout as a "normal" growatt datalogger record. 
The battery has a direct connection with growatt and you can see and set values via the growatt website and shinephone?

**Comment by johanmeijer** [comment]
Is it possible for you to take out the whole record checking part in grottproxy.py? This part: 

       #test if record is not corrupted
        vdata = "".join("{:02x}".format(n) for n in data)
        validatecc = validate_record(vdata)
        if validatecc != 0 : 
            print(f"\t - Grott - grottproxy - Invalid data record received, processing stopped for this record")
            #Create response if needed? 
            #self.send_queuereg[qname].put(response)
            return  

And the configure in grott.ini minrecl = 1. Hopefully grott will show more information then and the records still will be sent to Growatt.

**Reply by cgatnet** [reply]
Yes, I will try.

**Reply by cgatnet** [reply]
it worked. add the log file here now (gzipped).
[grott-test.log.gz](https://github.com/user-attachments/files/16427289/grott-test.log.gz)

**Reply by cgatnet** [reply]
after this, my noah went offline. Looks like I somehow lost the connection before rerouting it. Will reconnect in the evening. Too much sun actually. Need to collect it all! ;o)

**Reply by johanmeijer** [reply]
This is really a total different protocol(encryption?)/data layout. This will not be easy to debug, maybe we can find something about it on the internet.

**Reply by robertzaage** [reply]
> This is really a total different protocol(encryption?)/data layout. This will not be easy to debug, maybe we can find something about it on the internet.

My 800M-X is doing a TLS handshake and transmits encrypted data. I think they verify the fingerprint, as this is easily doable on ESP MCUs and it stuck on a handshake loop when I configured a MITM proxy.

**Comment by cgatnet** [comment]
This is what is shown on growatt webpage:
![Bildschirmfoto 2024-07-30 um 11 33 59](https://github.com/user-attachments/assets/b469c04a-a59b-42ae-adba-7a5626795666)

**Comment by PSMGoossens** [comment]
Hi,
It seems that growatt is using a MQTT server to communicate and not using Modbus TCP anymore.
I will see if I can intercept something :)

**Reply by robertzaage** [reply]
Only problem is the TLS encryption... I tried tearing down a 800M-X to dump the firmware, but the case was full of glue and I gave up.

**Reply by justinh998** [reply]
> Hi, It seems that growatt is using a MQTT server to communicate and not using Modbus TCP anymore. I will see if I can intercept something :)

Any News on that ?

**Comment by johanmeijer** [comment]
see also #552
